x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2867 @@
+
+_Source: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/rule_converters.ts)_
+
+```ts
+export const internalRuleToAPIResponse = (rule) => {
+  return {
+    ...commonParamsCamelToSnake(rule.params), // <--- rule params are converted here
+    ...typeSpecificCamelToSnake(rule.params),
+    // [... more object properties ...]
+  };
+};
+
+export const commonParamsCamelToSnake = (params: BaseRuleParams) => {
+  const { immutable, rule_source } = normalizeRuleSourceSchemaOnRuleRead(params);
+
+  return {
+    immutable,
+    rule_source,
+    // [... more object properties ...]
+  };
+};
+```
+
+And the `normalizeRuleSourceSchemaOnRuleRead` can be defined so:
+
+_Source: x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/prebuilt_rule_schema_migrations.ts_ (New file)
+
+```ts
+interface OnReadNormalizationResult {
+  immutable: IsRuleImmutable;
+  rule_source: RuleSource
+}
+
+/**
+ * To calculate `rule_source`:
+ * - Use `rule_source` field  if already exists in the internal rule object params.
+ * - If it does not exist, check the value `immutable` from the rules param:
+ *   - if it is `true`, create a new `rule_source` object, with `isCustomized` set to `false` and `repoNam` set to `elastic_prebuilt` (Use case of rule_source rules that have not been yet migrated-on-write)
+ *   - otherwise, the field should be `undefined` (use case of custom rules).
+ */
+const getRuleSourceValueForRuleRead = (ruleParams: BaseRuleParams): RuleSource | undefined => {
+  if (ruleParams.rule_source) {
+    return ruleParams.rule_source;
+  }
+
+  if (ruleParams.immutable) {
+    return {
+      type: 'external'
+      isCustomized: false,
+    };
+  }
+
+  return {
+    type: 'internal'
+  };
+};
+
+export const normalizePrebuiltSchemaOnRuleRead = (
+  ruleParams: BaseRuleParams
+): OnReadNormalizationResult => {
+
+/**
+ * To calculate `immutable`:
+ * - Checks if the `rule_source` field exists in the rule's parameters.
+ *   - If `rule_source` exists and its type is external, sets `immutable` to `true`. (Use case of Rules that have already been migrated-on-write)
+ *   - If `rule_source` does not exist, return the value of the params' `immutable` field. (Use case of Rules that have not yet been migrated on write.)
+ */
+  const isRuleSourceExternal = ruleParams.rule_source && ruleParams.rule_source.type === 'external';
+  const immutable = Boolean(isRuleSourceExternal) || Boolean(ruleParams.immutable);
+  const rule_source = getRuleSourceValueForRuleRead(ruleParams);
+
+  return {
+    immutable,
+    rule_source,
+  };
+};
+```
+
+##### Rule Management endpoints that will perform normalization-on-read
```

Comment:
nit: Having a table like

| API endpoint | migration type | comments |

could improve readability.

Reply:
++ I think it's a great idea

Reply:
Added a table and removed much of the text. I think it's much clearer now.